### PR TITLE
Read Committed 40001 errors; lock promotion

### PR DESCRIPTION
**Pull request by taroface**
DOC-9699
DOC-9600
DOC-9878
DOC-9869
DOC-9969
DOC-9621

- Remove Preview callout for Read Committed in 24.1
- Add enterprise license info
- Clarify that Read Committed txns don't require client-side retry handling
- Document lock promotion behavior (in SELECT FOR UPDATE/SHARE doc)
- Update description of Read Committed cluster setting behavior (true by default)
- Update various docs to more strongly incorporate Read Committed. **Question for reviewers:** Do the [Life of a Distributed Transaction](https://www.cockroachlabs.com/docs/v24.1/architecture/life-of-a-distributed-transaction.html) and [Transaction Layer](https://www.cockroachlabs.com/docs/v24.1/architecture/transaction-layer.html) (architecture) docs require more substantial updates? @nvanbenschoten @michae2 

The above updates are roughly separated by commit, in case it's easier to review that way.

### CODE REVIEW

File: src/current/v24.1/read-committed.md [hunk 1]
```diff
@@ -405,7 +405,7 @@ To use locking reads:
 
 - If you need to read and later update a row within a transaction, use `SELECT ... FOR UPDATE` to acquire an exclusive lock on the row. This guarantees data integrity between the transaction's read and write operations.
 
-- If you need to read the latest version of a row, but not update the row, use `SELECT ... FOR SHARE` to block all concurrent writes on the row without unnecessarily blocking concurrent reads.
+- If you need to read the latest version of a row, but not update the row, use `SELECT ... FOR SHARE` to acquire a shared lock on the row. This blocks all concurrent writes on the row without unnecessarily blocking concurrent reads.
```
Comment (nvanbenschoten) [inline src/current/v24.1/read-committed.md:397]: Consider adding "concurrent reads or other `SELECT ... FOR SHARE` queries".

File: src/current/v24.1/read-committed.md [hunk 2]
```diff
@@ -25,18 +25,7 @@ If your workload is already running well under `SERIALIZABLE` isolation, Cockroa
...
-~~~ sql
-SET CLUSTER SETTING sql.txn.read_committed_isolation.enabled = 'true';
-~~~
-
-After you enable the cluster setting, you can set `READ COMMITTED` as the [default isolation level](#set-the-default-isolation-level-to-read-committed) or [begin a transaction](#set-the-current-transaction-to-read-committed) as `READ COMMITTED`.
-
-{{site.data.alerts.callout_info}}
-If the cluster setting is not enabled, `READ COMMITTED` transactions will run as `SERIALIZABLE`.
-{{site.data.alerts.end}}
+By default, the `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}) is enabled. If the cluster setting is `false`, `READ COMMITTED` transactions will run as `SERIALIZABLE`.
```
Comment (nvanbenschoten) [inline src/current/v24.1/read-committed.md:28]: Do we want to talk somewhere about how a user can determine whether their application was previously asking for READ COMMITTED and getting SERIALIZABLE? This is now possible with the observability that Rafi added in https://github.com/cockroachdb/cockroach/pull/118497.

The thinking is that users will want to check this metric before performing the v24.1 upgrade to determine whether the upgrade will change the behavior of their application. For some users, that may be a welcomed change. For others, it could cause issues.

File: src/current/v24.1/demo-serializable.md [hunk 3]
```diff
@@ -545,5 +545,5 @@ Explore other core CockroachDB benefits and features:
 You might also want to learn more about how transactions work in CockroachDB and in general:
 
 - [Transactions Overview]({% link {{ page.version.version }}/transactions.md %})
-- [Real Transactions are Serializable](https://www.cockroachlabs.com/blog/acid-rain/)
```
Comment (nvanbenschoten) [inline src/current/v24.1/demo-serializable.md:548]: 👋

**Review by nvanbenschoten — APPROVED** [review]

File: src/current/v24.1/read-committed.md (hunk 1 above)
Comment (michae2) [reply src/current/v24.1/read-committed.md:397]: It might be worth qualifying the first sentence like:

> If you need to read the latest version of a row, and then update a _different_ row within a transaction, use ...

File: src/current/v24.1/transactions.md [hunk 4]
```diff
@@ -57,7 +57,7 @@ To handle errors in transactions, you should check for the following types of se
...
-**Transaction Retry Errors** | Errors with the code `40001` and string `restart transaction`, which indicate that a transaction failed because it could not be placed in a [serializable ordering]({% link {{ page.version.version }}/demo-serializable.md %}) of transactions by CockroachDB. For details on transaction retry errors and how to resolve them, see the [Transaction Retry Error Reference]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#actions-to-take).
+**Transaction Retry Errors** | Errors with the code `40001` and string `restart transaction`, which indicate that a transaction failed because it could not be placed in a [serializable ordering]({% link {{ page.version.version }}/demo-serializable.md %}) of transactions by CockroachDB. This occurs under [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}) isolation and only rarely under [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %})  isolation. For details on transaction retry errors and how to resolve them, see the [Transaction Retry Error Reference]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#actions-to-take).
```
Comment (michae2) [inline src/current/v24.1/transactions.md:60]: nit: two spaces before isolation

**Review by michae2 — APPROVED** [review]
These are really excellent docs!

File: src/current/v24.1/read-committed.md (hunk 2 above)
Comment (taroface) [reply src/current/v24.1/read-committed.md:28]: I'm going to add this in a separate PR as I need Rafi to review the docs for those metrics. Thanks for calling this out.

Comment (taroface) [reply src/current/v24.1/read-committed.md]: Actually just made another commit here. @rafiss (and anyone else) would you mind reviewing? https://github.com/cockroachdb/docs/pull/18522/commits/ab334f79c2560323d64f04a4a98db5691bbaa8ac

**Review by rafiss — APPROVED** [review]
"check whether transactions are being upgraded" lgtm!

File: src/current/_includes/v24.1/app/retry-errors.md [hunk 5]
```diff
@@ -1,3 +1,3 @@
 {{site.data.alerts.callout_info}}
-Your application should [use a retry loop to handle transaction errors]({% link {{ page.version.version }}/query-behavior-troubleshooting.md %}#transaction-retry-errors) that can occur under [contention]({{ link_prefix }}performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention).
+When running under the default [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}) isolation level, your application should [use a retry loop to handle transaction errors]({% link {{ page.version.version }}/query-behavior-troubleshooting.md %}#transaction-retry-errors) that can occur under [contention]({{ link_prefix }}performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention). Client-side retry handling is **not** necessary under [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %}) isolation.
```
Comment (mdlinville) [inline src/current/_includes/v24.1/app/retry-errors.md:2]: Is this the best link for `SERIALIZABLE`? I expected it to be a reference.

File: src/current/_includes/v24.1/faq/clock-synchronization-effects.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
...
 
-While [serializable consistency](https://wikipedia.org/wiki/Serializability) is maintained regardless of clock skew, skew outside the configured clock offset bounds can result in violations of single-key linearizability between causally dependent transactions. It's therefore important to prevent clocks from drifting too far by running [NTP](http://www.ntp.org/) or other clock synchronization software on each node.
+While [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}) and [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %}) transactions both serve globally consistent ("non-stale") reads and [commit atomically]({% link {{ page.version.version }}/developer-basics.md %}#how-transactions-work-in-cockroachdb) regardless of clock skew, skew outside the configured clock offset bounds can result in violations of single-key linearizability between causally dependent transactions. It's therefore important to prevent clocks from drifting too far by running [NTP](http://www.ntp.org/) or other clock synchronization software on each node.
```
Comment (mdlinville) [inline src/current/_includes/v24.1/faq/clock-synchronization-effects.md:3]: This is hard to follow. I know that you are just adding a link here, but I wonder if it could be re-ordered so that the sense of the last sentence goes first, or simplified somehow. I was lost by the time I got to "linearizability" and I'm not convinced this is a word (neither is Chrome spell check).

File: src/current/v24.1/frequently-asked-questions.md [hunk 7]
```diff
@@ -76,17 +76,17 @@ In a CockroachDB cluster spread across multiple geographic regions, the round-tr
...
 
 - Writes are serviced using the [Raft consensus algorithm](https://raft.github.io/), a popular alternative to <a href="https://www.microsoft.com/research/publication/paxos-made-simple/" data-proofer-ignore>Paxos</a>. A consensus algorithm guarantees that any majority of replicas together always agree on whether an update was committed successfully. Updates (writes) must reach a majority of replicas (2 out of 3 by default) before they are considered committed.
 
-  To ensure that a write transaction does not interfere with read transactions that start after it, CockroachDB also uses a [timestamp cache]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#timestamp-cache) which remembers when data was last read by ongoing transactions.
+- Stored data is versioned with MVCC, so under `SERIALIZABLE` isolation, [reads simply limit their scope to the data visible at the time the read transaction started]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#time-and-hybrid-logical-clocks).
```
Comment (mdlinville) [inline src/current/v24.1/frequently-asked-questions.md:85]: MVCC could link to https://www.cockroachlabs.com/docs/stable/architecture/storage-layer#mvcc

File: src/current/v24.1/frequently-asked-questions.md [hunk 8]
```diff
@@ -76,17 +76,17 @@ In a CockroachDB cluster spread across multiple geographic regions, the round-tr
...
+
+To ensure that a write transaction does not interfere with read transactions that start after it, CockroachDB also uses a [timestamp cache]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#timestamp-cache) which remembers when data was last read by ongoing transactions.
```
Comment (mdlinville) [inline src/current/v24.1/frequently-asked-questions.md:87]:
```suggestion
To ensure that a write transaction does not interfere with read transactions that start after it, CockroachDB also uses a [timestamp cache]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#timestamp-cache) which remembers when data was last read by an ongoing transaction.
```

File: src/current/v24.1/frequently-asked-questions.md [hunk 9]
```diff
@@ -76,17 +76,17 @@ In a CockroachDB cluster spread across multiple geographic regions, the round-tr
...
 
-  This ensures that clients always observe serializable consistency with regards to other concurrent transactions.
+This ensures that clients can always observe `SERIALIZABLE` consistency while issuing multiple concurrent transactions.
```
Comment (mdlinville) [inline src/current/v24.1/frequently-asked-questions.md:89]: Nit: This could be combined with the previous paragraph.

File: src/current/v24.1/migrate-from-oracle.md [hunk 10]
```diff
@@ -325,11 +325,11 @@ The last phase of the migration process is to change the [transactional behavior
...
+As CockroachDB does not allow serializable anomalies under `SERIALIZABLE` isolation, [transactions]({% link {{ page.version.version }}/begin-transaction.md %}) may experience deadlocks or [read/write contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention). This is expected during concurrency on the same keys. These can be addressed with either [automatic retries]({% link {{ page.version.version }}/transactions.md %}#automatic-retries) or [client-side transaction retry handling]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#client-side-retry-handling).
```
Comment (mdlinville) [inline src/current/v24.1/migrate-from-oracle.md:332]:
```suggestion
Because CockroachDB does not allow serializable anomalies under `SERIALIZABLE` isolation, [transactions]({% link {{ page.version.version }}/begin-transaction.md %}) may experience deadlocks or [read/write contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention). This is expected during concurrency on the same keys. These can be addressed with either [automatic retries]({% link {{ page.version.version }}/transactions.md %}#automatic-retries) or [client-side transaction retry handling]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#client-side-retry-handling).
```
"during concurrency" seems a bit strange.

File: src/current/v24.1/performance-best-practices-overview.md [hunk 11]
```diff
@@ -338,15 +338,15 @@ When the `transaction_rows_read_err` [session setting]({% link {{ page.version.v
...
 
-By default under [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}) isolation, transactions that operate on the same index key values (specifically, that operate on the same [column family]({% link {{ page.version.version }}/column-families.md %}) for a given index key) are strictly serialized to obey transaction isolation semantics. To maintain this isolation, writing transactions ["lock" rows]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#writing) to prevent interactions with concurrent transactions. 
+Writing transactions ["lock" rows]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#writing) to prevent interactions with concurrent transactions. Locking reads issued with [`SELECT ... FOR UPDATE`]({% link {{ page.version.version }}/select-for-update.md %}) perform a similar function by placing an [*exclusive lock*]({% link {{ page.version.version }}/select-for-update.md %}#lock-strengths) on rows, which can cause contention.
```
Comment (mdlinville) [inline src/current/v24.1/performance-best-practices-overview.md:341]: I think the first sentence is a fragment.

File: src/current/v24.1/performance-recipes.md [hunk 12]
```diff
@@ -136,7 +136,7 @@ These are indicators that transaction retries occurred in the past:
 
 Identify the transactions that are in conflict, and unblock them if possible. In general, take steps to [reduce transaction contention](#reduce-transaction-contention).
 
-In addition, implement [client-side retry handling]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#client-side-retry-handling) so that your application can respond to [transaction retry errors]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}) that are emitted when CockroachDB cannot [automatically retry]({% link {{ page.version.version }}/transactions.md %}#automatic-retries) a transaction.
+When running under `SERIALIZABLE` isolation, implement [client-side retry handling]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#client-side-retry-handling) so that your application can respond to [transaction retry errors]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}) that are emitted when CockroachDB cannot [automatically retry]({% link {{ page.version.version }}/transactions.md %}#automatic-retries) a transaction.
```
Comment (mdlinville) [inline src/current/v24.1/performance-recipes.md:139]: I think this is a fragment.

**Review by mdlinville — APPROVED** [review]
Just a few nits

File: src/current/_includes/v24.1/app/retry-errors.md (hunk 5 above)
Comment (taroface) [reply src/current/_includes/v24.1/app/retry-errors.md:2]: Yeah, we lack a dedicated "reference" doc since `SERIALIZABLE` was synonymous with "CockroachDB" until recently. The closest would be the Transaction Layer doc, but that one is too low-level to be an intro.

File: src/current/_includes/v24.1/faq/clock-synchronization-effects.md (hunk 6 above)
Comment (taroface) [reply src/current/_includes/v24.1/faq/clock-synchronization-effects.md:3]: Fixed I think!
